### PR TITLE
tackle-123-copy-assessment-between-2-apps

### PR DESCRIPTION
**Pull request by jonathanvila**
Issue : https://github.com/konveyor/tackle-pathfinder/issues/34

### Features covered
* New endpoint POST /assessments/copy/sourceApplicationId={appId}&targetApplicationId={appId} that will copy the assessment of the source application into the target application.

### Unit , Integration and End2End Tests cases covered
* Integration Test
  * given_ApplicationAssessed_When_CopyAssessmentToAnotherNonAssessedApp_Then_NewAssessmentIsCreatedForTargetApplicationWithSameValues
  * given_ApplicationAssessed_When_CopyAssessmentToAnotherAssessedApp_Then_BadRequestIsAssessed
  * given_ApplicationAssessedButDeleted_When_CopyAssessmentToAnotherAssessedApp_Then_404NotFoundIsReturned
  * given_ApplicationNotAssessed_When_CopyAssessmentToAnotherNotAssessedApp_Then_BadRequestIsAssessed
* Unit Test
  * given_AlreadyAssessedApplication_When_CopyAssessment_ShouldCopyAllUserFieldValues
* End2End Test on minikube with native image
  * Given an application assessed, copy the assessment to another application

### Pre steps Minikube: 
1. Install minikube : https://minikube.sigs.k8s.io/docs/start/
1. Start minikube : minikube start --kubernetes-version=v1.20.2 --cpus 4 --memory 12000
1. Enable ingress addons : minikube addons enable ingress
1. Create tackle namespace
```Shell
kubectl create namespace tackle
``` 
1. Deploy tackle ( to have the keycloak instance running ) on your K8s cluster
```Shell
$ kubectl apply -f  https://raw.githubusercontent.com/konveyor/tackle/main/kubernetes/kubernetes-tackle.yaml -n tackle
```
5. Edit the Ingress object to allow direct request to Pathfinder API
```Shell
$ kubectl edit ingress tackle -n tackle
```
Add this below `paths:`
```Yaml
     - backend:
          service:
            name: tackle-pathfinder
            port:
              number: 8080 
        path: /pathfinder
        pathType: ImplementationSpecific
```
6. Build and push 
```Shell
$ ./mvnw -U -B package -Dquarkus.container-image.push=true -Dquarkus.container-image.group={your quay user} -Dquarkus.container-image.registry=quay.io -Dquarkus.container-image.username={your quay user} -Dquarkus.container-image.password={your quay pwd} -Pnative
```
7. Change the image in the deployment tackle-pathfinder to be the one you have pushed
8. Rollout to update the image
```Shell
$ kubectl rollout restart deployment tackle-pathfinder -n tackle
```

### Test case with Minikube : Do the full API test
```Shell
$ .github/scripts/check_api.sh
```
> Expect 
```Shell
+++++ API CHECK SUCCESSFUL ++++++
```

### Pre steps non containerised local test:
1. run keycloak
```Shell
podman run -it --name keycloak --rm \     
            -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=admin -e KEYCLOAK_IMPORT=/tmp/keycloak/quarkus-realm.json \
            -e DB_VENDOR=h2 -p 8180:8080 -p 8543:8443 -v ./src/main/resources/keycloak:/tmp/keycloak:Z \
            jboss/keycloak:12.0.2
```
1. run postgresql
```Shell
podman run -it \                          
            --name postgres-pathfinder -e POSTGRES_USER=pathfinder \                                               
            -e POSTGRES_PASSWORD=pathfinder -e POSTGRES_DB=pathfinder_db \                              
            -p 5433:5432 postgres:10.6
```
1. run application
```Shell
./mvnw quarkus:dev
```
 
### Test case with local non containerised : Do the full API test
```Shell
$ .github/scripts/check_api.sh localhost:8085 localhost:8180
```
> Expect 
```Shell
+++++ API CHECK SUCCESSFUL ++++++
```

### NOTE 
Script Test step number involved in this feature : 13,14

### CODE REVIEW

**Comment by carlosthe19916** [comment]
@jonathanvila @m-brophy my two cents:
- For starting a copy the target and source data is being inserted through QueryParams. https://github.com/konveyor/tackle-pathfinder/blob/170dc8edb2ad3a62fcfeaccd9e29eb729c524b7d/src/main/java/io/tackle/pathfinder/controllers/AssessmentsResource.java#L75 
My opinion is that in POST/PUT operations we should be consistent with the rest of the endpoints and pass the data within the body and not through QueryParams. Technically nothing stops us from using QueryParams but I think we should have at least basic standards. The standard I propose is `use queryParams` if they are optional values,  `use the body or pathParams` if the data is mandatory. Wdyt? 
- For copying an assessment the current PR is using `targetApplicationId and sourceApplicationId` but I think what we should send to the backend is `targetAssessmentId and sourceApplicationIds` using `/POST /assessments/{targetAssessmentId}/copy` and within the body of the request we should send the list of applicationIds we want to copy the assessment. e.g:
```
POST /assessments/1
{
 applicationIdSource: [1,2,3,4,5,6]
}
```
Of course for the current PR we should only copy assessments one by one but for the future, I think we should copy multiple assessments.

**Review by m-brophy — CHANGES_REQUESTED** [review]

**Comment by jonathanvila** [comment]
>   My opinion is that in POST/PUT operations we should be consistent with the rest of the endpoints and pass the data within the body and not through QueryParams. Technically nothing stops us from using QueryParams but I think we should have at least basic standards. The standard I propose is `use queryParams` if they are optional values,  `use the body or pathParams` if the data is mandatory. Wdyt?

I would say Queryparams can be used when are not part of an entity to be stored. Optional/Mandatory should be the criteria to use query params in my opinion. 
Few criterias : 
* if you talk about 2,3 fields then Querystring could be a good place
* if you are sending DATA (not params) it should go in the body
* if you are sending parameters that configure an opperation, querystring could be a good place
* 
Here there are few links that can support current approach : 
https://stackoverflow.com/questions/25385559/rest-api-best-practices-args-in-query-string-vs-in-request-body
https://www.moesif.com/blog/technical/api-design/REST-API-Design-Best-Practices-for-Parameters-and-Query-String-Usage/

Obviously everything is opinionable in API design.

**Comment by jonathanvila** [comment]
> * For copying an assessment the current PR is using `sourceApplicationId and targetApplicationId` but I think what we should send to the backend is `sourceAssessmentId and targetApplicationIds` using `/POST /assessments/{sourceAssessmentId}/copy` and within the body of the request we should send the list of applicationIds we want to copy the assessment to. e.g:
> 
> ```
> POST /assessments/1/copy
> {
>  applicationIdTarget: [1,2,3,4,5,6]
> }
> ```
> 
> Of course for the current PR we should only copy assessments one by one but for the future, I think we should copy multiple assessments.

The current scope for this PR is copying one by one. There's another Ticket to cover the bulk copy, that will need a little discussion to cover ( https://github.com/konveyor/tackle-pathfinder/issues/38 ) : 
* sync or async , as it can take long to copy to several apps
* response as a list with status for each target app, as it could happen a subgroup of target apps fails , due to they already have an assessment, process exceptions, etc.

Appartfrom that the url you are proposing is copying an assessment NOT an application. If the application has more elements inside the operation is intended to copy everything not an assessment alone.
The right way to do it would be : /application/1234/copy

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java [hunk 1]
```diff
@@ -199,8 +200,98 @@ public AssessmentHeaderDto updateAssessment(@NotNull Long assessmentId, @NotNull
...
+        }
+
+        throw new BadRequestException();
+    }
+
+    @Transactional
+    private AssessmentQuestionnaire copyQuestionnaireBetweenAssessments(Assessment sourceAssessment, Assessment targetAssessment) {
+        AssessmentQuestionnaire questionnaire = AssessmentQuestionnaire.builder()
+                                                .assessment(targetAssessment)
+                                                .questionnaire(sourceAssessment.assessmentQuestionnaire.questionnaire)
```
[the remark this answers is not in the document]
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:255]: In fact I've seen there's no such field on the Questionnaire to store the assessmentQuestionnaires.

[the remark this answers is not in the document]
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:255]: So no need to add to questionnaire.

**Comment by m-brophy** [comment]
My comments have all been addressed satisfactorily so when you are happy @carlosthe19916 can you merge this PR?

**Pull request merged**